File: android/app/src/main/java/com/opbay/client/minecraft/Versions.kt
package com.opbay.client.minecraft

import com.opbay.client.data.Paths
import com.opbay.client.data.VersionChannel
import com.opbay.client.data.VersionSummary
import com.opbay.client.net.fetchJson
import com.opbay.client.net.json
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject

private const val MANIFEST_URL = "https://piston-meta.mojang.com/mc/game/version_manifest_v2.json"

@Serializable
data class Artifact(
    val path: String? = null,
    val sha1: String = "",
    val size: Long = 0,
    val url: String = ""
)

@Serializable
data class Rule(
    val action: String = "allow",
    val os: Os? = null,
    val features: Map<String, Boolean>? = null
) {
    @Serializable
    data class Os(val name: String? = null, val version: String? = null, val arch: String? = null)
}

@Serializable
data class Library(
    val name: String,
    val downloads: Downloads? = null,
    val url: String? = null,
    val natives: Map<String, String>? = null,
    val rules: List<Rule>? = null
) {
    @Serializable
    data class Downloads(
        val artifact: Artifact? = null,
        val classifiers: Map<String, Artifact>? = null
    )
}

@Serializable
data class AssetIndexRef(
    val id: String,
    val sha1: String = "",
    val size: Long = 0,
    val totalSize: Long = 0,
    val url: String = ""
)

@Serializable
data class VersionJson(
    val id: String,
    val inheritsFrom: String? = null,
    val type: String = "release",
    val mainClass: String = "",
    val minecraftArguments: String? = null,
    val arguments: Arguments? = null,
    val assetIndex: AssetIndexRef? = null,
    val assets: String? = null,
    val downloads: Map<String, Artifact>? = null,
    val libraries: List<Library> = emptyList(),
    val javaVersion: JavaVersion? = null
) {
    @Serializable
    data class JavaVersion(val component: String = "", val majorVersion: Int = 8)

    /**
     * Argument lists mix bare strings with conditional objects, so they stay as
     * raw JSON and are interpreted at launch time.
     */
    @Serializable
    data class Arguments(
        val game: List<JsonElement> = emptyList(),
        val jvm: List<JsonElement> = emptyList()
    )
}

@Serializable
private data class ManifestEntry(
    val id: String,
    val type: String,
    val url: String,
    val releaseTime: String = ""
)

@Serializable
private data class Manifest(
    val latest: Latest = Latest(),
    val versions: List<ManifestEntry> = emptyList()
) {
    @Serializable
    data class Latest(val release: String = "", val snapshot: String = "")
}

object Versions {

    private var cache: Pair<Long, Manifest>? = null

    private suspend fun manifest(): Manifest {
        cache?.let { (fetchedAt, data) ->
            if (System.currentTimeMillis() - fetchedAt < 10 * 60_000) return data
        }
        val fresh: Manifest = fetchJson(MANIFEST_URL)
        cache = System.currentTimeMillis() to fresh
        return fresh
    }

    /** Every published version, including the 2010-era alpha and beta builds. */
    suspend fun list(): List<VersionSummary> = manifest().versions.map { entry ->
        VersionSummary(
            id = entry.id,
            channel = when (entry.type) {
                "release" -> VersionChannel.RELEASE
                "snapshot" -> VersionChannel.SNAPSHOT
                "old_beta" -> VersionChannel.OLD_BETA
                else -> VersionChannel.OLD_ALPHA
            },
            releaseTime = entry.releaseTime,
            url = entry.url
        )
    }

    suspend fun latestRelease(): String = manifest().latest.release

    /** Reads a version json from disk, fetching it from Mojang on first use. */
    suspend fun load(paths: Paths, versionId: String): VersionJson = withContext(Dispatchers.IO) {
        val file = paths.versionJson(versionId)
        if (file.isFile) {
            runCatching { json.decodeFromString<VersionJson>(file.readText()) }
                .getOrNull()
                ?.let { return@withContext it }
        }

        val entry = manifest().versions.firstOrNull { it.id == versionId }
            ?: throw IllegalArgumentException("Minecraft sürümü bulunamadı: $versionId")

        val text = fetchJson<JsonObject>(entry.url).toString()
        file.parentFile?.mkdirs()
        file.writeText(text)
        json.decodeFromString(text)
    }

    /**
     * Flattens a version and everything it inherits from. Mod loaders publish
     * partial version files that build on the vanilla one.
     */
    suspend fun resolve(paths: Paths, versionId: String): VersionJson {
        val chain = mutableListOf<VersionJson>()
        val seen = mutableSetOf<String>()
        var current: String? = versionId

        while (current != null && seen.add(current)) {
            val version = load(paths, current)
            chain += version
            current = version.inheritsFrom
        }

        // Parents first, so children override them.
        return chain.reversed().reduce { parent, child ->
            child.copy(
                libraries = parent.libraries + child.libraries,
                arguments = VersionJson.Arguments(
                    game = (parent.arguments?.game ?: emptyList()) + (child.arguments?.game ?: emptyList()),
                    jvm = (parent.arguments?.jvm ?: emptyList()) + (child.arguments?.jvm ?: emptyList())
                ),
                // A child that omits these must keep the parent's values.
                assetIndex = child.assetIndex ?: parent.assetIndex,
                assets = child.assets ?: parent.assets,
                downloads = child.downloads ?: parent.downloads,
                javaVersion = child.javaVersion ?: parent.javaVersion,
                minecraftArguments = child.minecraftArguments ?: parent.minecraftArguments,
                mainClass = child.mainClass.ifEmpty { parent.mainClass }
            )
        }
    }
}

/**
 * Android is not one of Mojang's target platforms, so rules are evaluated as if
 * this were Linux — which is what the Java runtime actually reports.
 */
object Rules {

    const val OS_NAME = "linux"

    val arch: String
        get() = when {
            android.os.Build.SUPPORTED_64_BIT_ABIS.any { it.startsWith("arm") } -> "aarch64"
            android.os.Build.SUPPORTED_64_BIT_ABIS.isNotEmpty() -> "x86_64"
            else -> "x86"
        }

    /** Rules are ordered and the last match wins; an empty list means allow. */
    fun allows(rules: List<Rule>?, features: Map<String, Boolean> = emptyMap()): Boolean {
        if (rules.isNullOrEmpty()) return true

        var allowed = false
        for (rule in rules) {
            var matches = true
            rule.os?.let { os ->
                if (os.name != null && os.name != OS_NAME) matches = false
                // Mojang writes x86/x86_64/arm64; treat anything unexpected as a mismatch.
                if (os.arch != null && os.arch != arch && !(os.arch == "x86_64" && arch == "aarch64")) {
                    matches = false
                }
            }
            rule.features?.forEach { (feature, expected) ->
                if ((features[feature] ?: false) != expected) matches = false
            }
            if (matches) allowed = rule.action == "allow"
        }
        return allowed
    }

    /** Expands the `${placeholder}` tokens used throughout argument templates. */
    fun substitute(template: String, values: Map<String, String>): String {
        var result = template
        for ((key, value) in values) result = result.replace("\${$key}", value)
        return result
    }

    /** Flattens Mojang's mixed string/conditional argument list. */
    fun flatten(
        entries: List<JsonElement>,
        values: Map<String, String>,
        features: Map<String, Boolean>
    ): List<String> {
        val output = mutableListOf<String>()
        for (entry in entries) {
            if (entry is JsonPrimitive) {
                output += substitute(entry.content, values)
                continue
            }

            val obj = entry.jsonObject
            val rules = obj["rules"]?.let { element ->
                runCatching { json.decodeFromJsonElement<List<Rule>>(element) }.getOrNull()
            }
            if (!allows(rules, features)) continue

            when (val value = obj["value"]) {
                is JsonPrimitive -> output += substitute(value.content, values)
                is JsonArray -> value.forEach { item ->
                    (item as? JsonPrimitive)?.let { output += substitute(it.content, values) }
                }
                else -> Unit
            }
        }
        return output
    }
}

File: src/main/minecraft/versions.ts
import fsp from 'node:fs/promises'
import path from 'node:path'
import type { VersionSummary } from '../../shared/types'
import { fetchJson } from './downloader'

const MANIFEST_URL = 'https://piston-meta.mojang.com/mc/game/version_manifest_v2.json'

export interface Rule {
  action: 'allow' | 'disallow'
  os?: { name?: string; version?: string; arch?: string }
  features?: Record<string, boolean>
}

export interface Artifact {
  path?: string
  sha1: string
  size: number
  url: string
}

export interface Library {
  name: string
  downloads?: { artifact?: Artifact; classifiers?: Record<string, Artifact> }
  url?: string
  /**
   * Marks a jar the launcher produced on this machine rather than fetched.
   *
   * OptiFine's library is patched out of the vanilla client jar at install time
   * and exists on no maven repository, so it has to stay on the classpath while
   * never being queued for download.
   */
  local?: boolean
  natives?: Record<string, string>
  extract?: { exclude?: string[] }
  rules?: Rule[]
}

export interface VersionJson {
  id: string
  inheritsFrom?: string
  type: string
  mainClass: string
  minecraftArguments?: string
  arguments?: { game?: (string | { rules: Rule[]; value: string | string[] })[]; jvm?: (string | { rules: Rule[]; value: string | string[] })[] }
  assetIndex?: { id: string; sha1: string; size: number; totalSize: number; url: string }
  assets?: string
  downloads?: Record<string, Artifact>
  libraries: Library[]
  javaVersion?: { component: string; majorVersion: number }
  logging?: { client?: { argument: string; file: { id: string; sha1: string; size: number; url: string }; type: string } }
}

interface ManifestEntry {
  id: string
  type: VersionSummary['type']
  url: string
  time: string
  releaseTime: string
  sha1: string
}

interface Manifest {
  latest: { release: string; snapshot: string }
  versions: ManifestEntry[]
}

let manifestCache: { fetchedAt: number; data: Manifest } | null = null

async function getManifest(): Promise<Manifest> {
  if (manifestCache && Date.now() - manifestCache.fetchedAt < 10 * 60_000) return manifestCache.data
  const data = await fetchJson<Manifest>(MANIFEST_URL)
  manifestCache = { fetchedAt: Date.now(), data }
  return data
}

export async function listVersions(): Promise<VersionSummary[]> {
  const manifest = await getManifest()
  return manifest.versions.map((entry) => ({
    id: entry.id,
    type: entry.type,
    releaseTime: entry.releaseTime
  }))
}

export async function latestRelease(): Promise<string> {
  return (await getManifest()).latest.release
}

/** Reads a version json from disk, downloading it from Mojang on first use. */
export async function loadVersionJson(
  dataDir: string,
  versionId: string,
  offline = false
): Promise<VersionJson> {
  const file = path.join(dataDir, 'versions', versionId, `${versionId}.json`)
  try {
    return JSON.parse(await fsp.readFile(file, 'utf8')) as VersionJson
  } catch {
    // Not cached yet — fall through to the manifest lookup below.
  }

  if (offline) {
    throw new Error(
      `Minecraft ${versionId} sürüm bilgisi bu bilgisayarda hazır değil. ` +
        'İnternete bağlanıp “Dosyaları önceden indir” işlemini çalıştırın.'
    )
  }

  const manifest = await getManifest()
  const entry = manifest.versions.find((candidate) => candidate.id === versionId)
  if (!entry) throw new Error(`Minecraft sürümü bulunamadı: ${versionId}`)

  const json = await fetchJson<VersionJson>(entry.url)
  await fsp.mkdir(path.dirname(file), { recursive: true })
  await fsp.writeFile(file, JSON.stringify(json, null, 2))
  return json
}

/**
 * Flattens a version and every parent it inherits from (mod loaders publish
 * partial version files that build on the vanilla one).
 */
export async function resolveVersion(
  dataDir: string,
  versionId: string,
  offline = false
): Promise<VersionJson> {
  const chain: VersionJson[] = []
  const seen = new Set<string>()
  let current: string | undefined = versionId

  while (current && !seen.has(current)) {
    seen.add(current)
    const json: VersionJson = await loadVersionJson(dataDir, current, offline)
    chain.push(json)
    current = json.inheritsFrom
  }

  // Walk parents first so children override them. Libraries are the exception:
  // the loader's entries go in front, because resolution keeps the first match
  // for a coordinate and a loader that replaces a vanilla library must win.
  return chain.reverse().reduce((merged, json) => ({
    ...merged,
    ...json,
    libraries: [...(json.libraries ?? []), ...(merged.libraries ?? [])],
    arguments: {
      game: [...(merged.arguments?.game ?? []), ...(json.arguments?.game ?? [])],
      jvm: [...(merged.arguments?.jvm ?? []), ...(json.arguments?.jvm ?? [])]
    },
    // A child that omits these must keep the parent's values.
    assetIndex: json.assetIndex ?? merged.assetIndex,
    assets: json.assets ?? merged.assets,
    downloads: json.downloads ?? merged.downloads,
    javaVersion: json.javaVersion ?? merged.javaVersion,
    minecraftArguments: json.minecraftArguments ?? merged.minecraftArguments
  })) as VersionJson
}

/** Compares two Minecraft versions well enough to sort a version list. */
export function compareVersions(a: string, b: string): number {
  const parse = (value: string): number[] => value.split(/[.\-+_]/).map((part) => Number.parseInt(part, 10) || 0)
  const left = parse(a)
  const right = parse(b)
  for (let i = 0; i < Math.max(left.length, right.length); i++) {
    const diff = (right[i] ?? 0) - (left[i] ?? 0)
    if (diff !== 0) return diff
  }
  return 0
}
